gtsfm/frontend/matcher/twoway_matcher.py

```diff
@@ -60,28 +74,54 @@ def match(
             distance_metric = cv.NORM_HAMMING
         else:
             raise NotImplementedError("The distance type is not in MatchingDistanceType")
-
         if descriptors_i1.size == 0 or descriptors_i2.size == 0:
             return np.array([])
 
-        # we will have to remove NaNs by ourselves
+        # Remove NaNs.
         valid_idx_i1 = np.nonzero(~(np.isnan(descriptors_i1).any(axis=1)))[0]
         valid_idx_i2 = np.nonzero(~(np.isnan(descriptors_i2).any(axis=1)))[0]
 
         descriptors_1 = descriptors_i1[valid_idx_i1]
         descriptors_2 = descriptors_i2[valid_idx_i2]
 
-        # run OpenCV's matcher
-        bf = cv.BFMatcher(normType=distance_metric, crossCheck=True)
-        matches = bf.match(descriptors_1, descriptors_2)
-        matches = sorted(matches, key=lambda r: r.distance)
-
+        # Run KNN matcher both ways.
+        bf_matcher = cv.BFMatcher(normType=distance_metric)
+        init_matches1 = bf_matcher.knnMatch(descriptors_1, descriptors_2, k=2)
+        init_matches2 = bf_matcher.knnMatch(descriptors_2, descriptors_1, k=2)
+
+        # Filter putative matches using cross check and Lowe's ratio test.
+        good_matches = []
+        for i in range(len(init_matches1)):
+            cond = True
+            if cross_check:
+                cond1 = cross_check and init_matches2[init_matches1[i][0].trainIdx][0].trainIdx == i
+                cond = cond and cond1
+            if ratio is not None and ratio < 1:
```

Comment:
very nice. could you add a unit test with say 3 toy 1-dimensional descriptors (i.e. on a number line) to make sure the ratio test is computed correctly?

Reply:
Yes, good idea.